libs/ts-rest/fastify/src/lib/ts-rest-fastify.ts

```diff
@@ -0,0 +1,237 @@
+  validateResponse,
+  Without,
+  ZodInferOrType,
+} from '@ts-rest/core';
+import * as fastify from 'fastify';
+
+type AppRouteQueryImplementation<T extends AppRouteQuery> = (
+  input: Without<
+    {
+      params: PathParamsWithCustomValidators<T>;
+      query: ZodInferOrType<T['query']>;
+      headers: LowercaseKeys<ZodInferOrType<T['headers']>> &
+        fastify.FastifyRequest['headers'];
+      request: fastify.FastifyRequest;
+      reply: fastify.FastifyReply;
+    },
+    never
+  >
+) => Promise<ApiRouteServerResponse<T['responses']>>;
+
+type WithoutFileIfMultiPart<T extends AppRouteMutation> =
+  T['contentType'] extends 'multipart/form-data'
+    ? Without<ZodInferOrType<T['body']>, File>
+    : ZodInferOrType<T['body']>;
+
+type AppRouteMutationImplementation<T extends AppRouteMutation> = (
+  input: Without<
+    {
+      params: PathParamsWithCustomValidators<T>;
+      query: ZodInferOrType<T['query']>;
+      body: WithoutFileIfMultiPart<T>;
+      headers: LowercaseKeys<ZodInferOrType<T['headers']>> &
+        fastify.FastifyRequest['headers'];
+      request: fastify.FastifyRequest;
+      reply: fastify.FastifyReply;
+    },
+    never
+  >
+) => Promise<ApiRouteServerResponse<T['responses']>>;
+
+type AppRouteImplementation<T extends AppRoute> = T extends AppRouteMutation
+  ? AppRouteMutationImplementation<T>
+  : T extends AppRouteQuery
+  ? AppRouteQueryImplementation<T>
+  : never;
+
+type RecursiveRouterObj<T extends AppRouter> = {
+  [TKey in keyof T]: T[TKey] extends AppRouter
+    ? RecursiveRouterObj<T[TKey]>
+    : T[TKey] extends AppRoute
+    ? AppRouteImplementation<T[TKey]>
+    : never;
+};
+
+type RegisterRouterOptions = {
+  logInitialization?: boolean;
+  jsonQuery?: boolean;
+  responseValidation?: boolean;
+};
+
+const validateRequest = (
+  request: fastify.FastifyRequest,
+  reply: fastify.FastifyReply,
+  schema: AppRouteQuery | AppRouteMutation,
+  options: RegisterRouterOptions
+) => {
+  const paramsResult = checkZodSchema(request.params, schema.pathParams, {
+    passThroughExtraKeys: true,
+  });
+
+  const headersResult = checkZodSchema(request.headers, schema.headers, {
+    passThroughExtraKeys: true,
+  });
+
+  const queryResult = checkZodSchema(
+    options.jsonQuery
+      ? parseJsonQueryObject(request.query as Record<string, string>)
+      : request.query,
+    schema.query
+  );
+
+  const bodyResult = checkZodSchema(
+    request.body,
+    'body' in schema ? schema.body : null
+  );
+
+  if (
+    !paramsResult.success ||
+    !headersResult.success ||
+    !queryResult.success ||
+    !bodyResult.success
+  ) {
+    reply.status(400).send({
+      queryParameterErrors: queryResult.success ? null : queryResult.error,
+      pathParameterErrors: paramsResult.success ? null : paramsResult.error,
+      headerErrors: headersResult.success ? null : headersResult.error,
+      bodyErrors: bodyResult.success ? null : bodyResult.error,
+    });
+
+    return null;
+  }
```

Comment:
This is great! This also made me think of how there is some need to customize these errors. What do you think that instead of setting the response here we instead throw a custom error `TsRestValidationError` that contains all of these error parameters.

We also provide a standard error handler to be used by the developer like this `fastify.setErrorHandler(TsRestErrorHandler)` that takes care returning the standard format you have here, but can be swapped out at will